mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,181 @@
+use crate::sync::Arc;
+use bytes::Bytes;
+use futures::pin_mut;
+use futures::StreamExt;
+use mountpoint_s3_client::types::ETag;
+use mountpoint_s3_client::ObjectClient;
+use parquet::errors::ParquetError;
+use parquet::file::footer::decode_footer;
+use parquet::file::metadata::ParquetMetaData;
+use parquet::file::reader::{ChunkReader, Length};
+use std::collections::HashMap;
+use std::io::{Read, Seek, SeekFrom};
+use tracing::debug_span;
+use tracing::trace;
+
+use super::PrefetchReadError;
+
+pub async fn read_parquet_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<Bytes, PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    let chunk_reader = ParquetS3ChunkReader {
+        client,
+        bucket: bucket.to_owned(),
+        key: key.to_owned(),
+        etag: if_match.clone(),
+        size: total_size,
+    };
+
+    parse_metadata_raw(&chunk_reader).map_err(|_| PrefetchReadError::GetRequestTerminatedUnexpectedly)
+}
+
+pub fn parse_byte_ranges(metadata: &ParquetMetaData) -> HashMap<(usize, usize), (u64, u64)> {
+    let mut byte_ranges = HashMap::new();
+
+    for (rowgroup_index, rowgroup) in metadata.row_groups().iter().enumerate() {
+        for (column_index, column_metadata) in rowgroup.columns().iter().enumerate() {
+            let start_byte = column_metadata.file_offset() as u64;
+            let end_byte = start_byte + column_metadata.compressed_size() as u64 - 1;
+
+            byte_ranges.insert((rowgroup_index, column_index), (start_byte, end_byte));
+        }
+    }
+
+    byte_ranges
+}
+
+fn parse_metadata_raw<R: ChunkReader>(chunk_reader: &R) -> Result<Bytes, ParquetError> {
+    let parquet_magic_len = 8; // Represents the 4 bit "size of metadata" + 4 bit "magic number" (https://parquet.apache.org/docs/file-format/)
+                               // check file is large enough to hold footer
+    let file_size = chunk_reader.len();
+    if file_size < (parquet_magic_len as u64) {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Size is smaller than footer".to_string(),
+        ));
+    }
+
+    let mut footer = [0_u8; 8];
+    chunk_reader.get_read(file_size - 8, 8)?.read_exact(&mut footer)?;
```

Comment:
it seems like we know exactly what range in the object we need after receiving this right?

Reply:
From that information, we'll get the last 8 bytes, and from that we can decode the size of the entire metadata